test/e2e/remediation_test.go

```diff
@@ -0,0 +1,477 @@
+	newM3DataTemplate.Spec.NetworkData = m3dataTemplate.Spec.NetworkData
+	newM3DataTemplate.Spec.ClusterName = clusterName
+	newM3DataTemplate.Spec.TemplateReference = m3dataTemplateName
+
+	newM3DataTemplate.ObjectMeta.Name = newM3dataTemplateName
+	newM3DataTemplate.ObjectMeta.Namespace = m3dataTemplate.Namespace
+
+	err = bootstrapClient.Create(ctx, newM3DataTemplate)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Creating a new Metal3MachineTemplate")
+	m3machineTemplate := capm3.Metal3MachineTemplate{}
+	m3machineTemplateName := fmt.Sprintf("%s-workers", clusterName)
+	Expect(bootstrapClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: m3machineTemplateName}, &m3machineTemplate)).To(Succeed())
+	newM3MachineTemplateName := "test-new-m3mt"
+
+	newM3MachineTemplate := m3machineTemplate.DeepCopy()
+	cleanObjectMeta(&newM3MachineTemplate.ObjectMeta)
+
+	newM3MachineTemplate.Spec.Template.Spec.Image = m3machineTemplate.Spec.Template.Spec.Image
+	newM3MachineTemplate.Spec.Template.Spec.DataTemplate.Name = newM3dataTemplateName
+	newM3MachineTemplate.ObjectMeta.Name = newM3MachineTemplateName
+
+	Expect(bootstrapClient.Create(ctx, newM3MachineTemplate)).To(Succeed(), "Failed to create new M3MachineTemplate")
+
+	By("Pointing MachineDeployment to the new Metal3MachineTemplate")
+	deployment := clusterv1.MachineDeployment{}
+	Expect(bootstrapClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: clusterName}, &deployment)).To(Succeed())
+
+	helper, err := patch.NewHelper(&deployment, bootstrapClient)
+	Expect(err).NotTo(HaveOccurred())
+
+	deployment.Spec.Template.Spec.InfrastructureRef = v1.ObjectReference{
+		Kind:       "Metal3MachineTemplate",
+		APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha5",
+		Name:       newM3MachineTemplateName,
+	}
+	deployment.Spec.Strategy.RollingUpdate.MaxUnavailable = &intstr.IntOrString{IntVal: 1}
+	Expect(helper.Patch(ctx, &deployment)).To(Succeed())
+
+	By("Waiting for 2 BMHs to be Ready")
+	Eventually(func(g Gomega) error {
+		bmhs := getAllBmhs(ctx, bootstrapClient, namespace, specName)
+		filtered := filterBmhsByProvisioningState(bmhs, bmh.StateReady)
+		Logf("There are %d BMHs in state %s", len(filtered), bmh.StateReady)
+		g.Expect(filtered).To(HaveLen(2))
+		return nil
+	}, e2eConfig.GetIntervals(specName, "wait-machine-remediation")...).Should(Succeed())
+
+	By("Waiting for 1 Metal3Data to refer to the old template")
+	Eventually(func(g Gomega) {
+		datas := capm3.Metal3DataList{}
+		g.Expect(bootstrapClient.List(ctx, &datas, client.InNamespace(namespace))).To(Succeed())
+		filtered := filterM3DataByReference(datas.Items, m3dataTemplateName)
+		g.Expect(filtered).To(HaveLen(1))
+	}, e2eConfig.GetIntervals(specName, "wait-deployment")...).Should(Succeed())
+
+	By("Scaling up KCP to 3 replicas")
+	scaleControlPlane(ctx, bootstrapClient, client.ObjectKey{Namespace: "metal3", Name: "test1"}, 3)
+
+	Byf("Waiting for all (%d) BMHs to be Provisioned", allMachinesCount-1)
```

Comment:
any reason we would wait for `allMachinesCount-1` here where we expect `allMachinesCount`?

Reply:
It was a bug. Good catch!